docs/routing.md

```diff
@@ -0,0 +1,248 @@
+
+If the current phase is `miss`, it will update the status code to 404 if there is no file found in the build output. Afterwards, if a file was found in the output, or the current phase is `error`, it moves on to the `hit` phase. It is important that the hit phase is run every time a match is found in the build output or when the current phase is `miss` or `error`, because this allows us to update the headers for the response with the current matched path.
+
+In all other cases, it attempts to retrieve the next phase, based on the current phase. Afterwards, it recursively calls the function so that it can check the next phase.
+
+#### Diagram
+
+```mermaid
+flowchart TD
+    checkPhase --for each route--> checkRoute
+    checkRoute --> CheckRouteResult{check result}
+    CheckRouteResult -->|error| checkPhase
+    CheckRouteResult -->|done| IsPhaseHit{phase == hit}
+    CheckRouteResult -->|skip or next| CheckRouteResult
+    IsPhaseHit --> |yes| EndIsPhaseHit[matching done]
+    IsPhaseHit --> |no| IsPhaseMiss{phase == miss}
+    IsPhaseMiss --> |yes| Set404[set 404]
+    Set404 --> NextPhaseIsWhat{path found or phase == miss/error}
+    IsPhaseMiss --> |no| NextPhaseIsWhat
+    NextPhaseIsWhat --> |true - set next phase to hit| checkPhase
+    NextPhaseIsWhat --> |get next phase| getNextPhase --> checkPhase
+
+    getNextPhase --> NextPhaseIs{determine next phase}
+    NextPhaseIs --> |none| NextIsFS[filesystem]
+    NextPhaseIs --> |filesystem| NextIsRW[rewrite]
+    NextPhaseIs --> |rewrite| NextIsRS[resource]
+    NextPhaseIs --> |resource| NextIsMS[miss]
+    NextPhaseIs --> |all others| NextIsMS
+```
+
+### Checking Source Routes
+
+For every source route in a phase, we need to check for a match and process it accordingly.
+
+#### Source Matching
+
+The first step in this part of the process is checking the value for the route matches up with the current path. At the same time, this also extracts any dynamic parameters.
+
+Checking a route is a match involves verifying that the request method is the same, the correct fields are or aren't specified, and the current path name matches.
+
+If the route was not a match, we simply return that we want to skip it.
+
+#### Overrides
+
+In different phases, different things may be applied to our response object, or modify the current requests details. The first thing we check for is whether a source route wants to override the current headers and status code and, if so, reset them.
+
+#### Middleware
+
+Next up, any middleware is run. Middleware occurs during the `none` phase, at the very start of the routing process. The response headers from the middleware and taken and applied to the request and the response object, where appropriate.
+
+Additionally, it is important to check whether the middleware was a rewrite, so that the current path can be updated.
+
+If running the middleware failed, we have to bail out of routing and return an error to the client.
```

Comment:
do we return an error response right await of do we go to the `error` phase?

Reply:
we return an error response back to the router, and the router then enters the error phase. will clarify that we enter the error phase right away.